File: Libs/pollilib/python/polliLib/stt.py
from __future__ import annotations

from typing import Any, Dict, Optional


class STTMixin:
    def transcribe_audio(
        self,
        audio_path: str,
        *,
        question: str = "Transcribe this audio",
        model: str = "openai-audio",
        provider: str = "openai",
        referrer: Optional[str] = None,
        token: Optional[str] = None,
        timeout: Optional[float] = 120.0,
    ) -> Optional[str]:
        import os, base64
        if not os.path.exists(audio_path):
            raise FileNotFoundError(audio_path)
        ext = os.path.splitext(audio_path)[1].lower().lstrip(".")
        if ext not in {"mp3", "wav"}:
            return None
        with open(audio_path, "rb") as f:
            b64 = base64.b64encode(f.read()).decode("utf-8")
        payload: Dict[str, Any] = {
            "model": model,
            "messages": [
                {
                    "role": "user",
                    "content": [
                        {"type": "text", "text": question},
                        {"type": "input_audio", "input_audio": {"data": b64, "format": ext}},
                    ],
                }
            ],
        }
        if referrer:
            payload["referrer"] = referrer
        if token:
            payload["token"] = token
        payload["safe"] = False
        url = f"{self.text_prompt_base}/{provider}"
        headers = {"Content-Type": "application/json"}
        resp = self.session.post(url, headers=headers, json=payload, timeout=timeout or self.timeout)
        resp.raise_for_status()
        data = resp.json()
        return data.get("choices", [{}])[0].get("message", {}).get("content")

File: Libs/pollilib/javascript/polliLib/stt.js
import fs from 'node:fs';

export const STTMixin = (Base) => class extends Base {
  async transcribe_audio(audioPath, { question = 'Transcribe this audio', model = 'openai-audio', provider = 'openai', referrer = null, token = null, timeoutMs = 120_000 } = {}) {
    if (!fs.existsSync(audioPath)) throw new Error(`File not found: ${audioPath}`);
    const ext = String(audioPath).split('.').pop().toLowerCase();
    if (!['mp3','wav'].includes(ext)) return null;
    const data = await fs.promises.readFile(audioPath);
    const b64 = data.toString('base64');
    const payload = {
      model,
      messages: [
        { role: 'user', content: [ { type: 'text', text: question }, { type: 'input_audio', input_audio: { data: b64, format: ext } } ] }
      ]
    };
    if (referrer) payload.referrer = referrer;
    if (token) payload.token = token;
    payload.safe = false;
    const url = `${this.textPromptBase}/${provider}`;
    const controller = new AbortController();
    const t = setTimeout(() => controller.abort(), timeoutMs || this.timeoutMs);
    try {
      const resp = await this.fetch(url, { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(payload), signal: controller.signal });
      if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
      const json = await resp.json();
      return json?.choices?.[0]?.message?.content;
    } finally { clearTimeout(t); }
  }
};

